feat(editor/subscription): add supervision indication

danyill opened the pull request:
(This PR has the same content as https://github.com/openscd/open-scd/pull/1061 but that was unfortunately closed/damaged, probably by a force-push on the PR for LGOS/LSVS without recognising the impact on a dependent branch.

It was not _very_ easy to recover and was done by producing a patch file for the squashed original branch this work was done in and then manually applying this on top of main -- line by line because I am paranoid -- and finally regenerating the snapshots. I hope to have done it correctly and not to have to do it again any time soon :sweat: )

1. I've augmented the message binding test files with LGOS and LSVS supervision nodes.
2. I've done what I can with the unit tests.
3. The logic of the getFirstSubscribedExtRef function is not very clean and I would be happy to take suggestions.

Jakob may have already reviewed this.

JakobVogelsang commented on src/editors/subscription/foundation.ts:
```diff
@@ -432,6 +433,55 @@ export function findOrCreateAvailableLNInst(
+ * @returns The supervision LN instance or null if not found
+ */
+export function getExistingSupervision(
+  extRef: Element | null,
```
I also would propose to simplify the logic a bit more. You need only the subscriber IED as input then you do not need to check for the first `ExtRef` and the control block or even simpler the `srcxxx` attributes that are nothing else then the object reference you are searching for in the subscriber IED.

danyill replied:
>Here we can use the srcxxx element and the serviceType. With that, the code could be much simpler.

> You need only the subscriber IED as input then you do not need to check for the first ExtRef and the control block or even simpler the srcxxx attributes that are nothing else then the object reference you are searching for in the subscriber IED.

Thank you for this comment. 
That seems like a good idea, I can refactor as suggested on the "Ed 2" assumption. I don't have a counter-example.

danyill replied:
I think your suggestions are excellent :100: and I have made some changes.
However, I would like to confirm one aspect. 

For the message binding plugin I think I do need the `getFirstSubscribedExtRef` function - are you suggesting this is unnecessary? I have to work with a source IED, a control block, and a sink IED and I must find the first `ExtRef` to be able to check it.

JakobVogelsang replied:
Well, I think that the control block and the sink IED is already all the information you need, because what you ultimately want to know is: whether there is a control block reference somewhere within a psecific DAI/DOI and LGOS/LSVS and sink IED.

The reason I think that is enough is that you are building an indication, after all. With a subscription supervision, you might want to go all the way over the `ExtRef` to be sure.

JakobVogelsang replied:
BUT what you have done works very well has no performance issues and it reads well. So it is up to you if you want to further simplify it :)

danyill replied:
I think I will merge for now but perhaps we can discuss some time and I would be happy to do a follow-up PR to simplify further.

JakobVogelsang reviewed (commented):
I think the general approach can be simplified. But it is debatable. Maybe I am missing something e.g. some vendor not having the `srcxxx` in external references

danyill commented on src/editors/subscription/foundation.ts:
```diff
@@ -432,6 +433,55 @@ export function findOrCreateAvailableLNInst(
+  subscribingIed: Element
+): Element | null {
+  const publishingIed = publishedControlBlock.closest('IED')!;
+  const dataSet = publishingIed.querySelector(
```
I will refactor it anyway, but for my education, what is dangerous about this query?

JakobVogelsang replied:
The name attribute of the `DataSet` element is unique within the parent anyLn. That means there can be two `DataSet`s having the same name within the same IED and you by mistake get the wrong one.

JakobVogelsang commented:
Sorry the trouble restoring your work in this PR @danyill . For the next time I found interactive rebase to be very helpful in this particular situation. This allows you to put your work on top of a new main. The interactive in particular is useful because we squash and merge. This means that a PR having say 10 commits is only one commit in main. In this case you can easily drop (d) all commits in the interactive rebase.

danyill commented:
> Sorry the trouble restoring your work in this PR @danyill . For the next time I found interactive rebase to be very helpful in this particular situation. This allows you to put your work on top of a new main. The interactive in particular is useful 

I'm not sure this is possible if the initial commit no longer exists, which is what I think happened here. I based on a commit that was removed through a force push.

However I resolved basically as you suggested, I rebased and squashed all commits that were "mine" and then manually applied it. I couldn't automatically apply it because the initial commit no longer existed if I understand correctly.

JakobVogelsang reviewed (approved):
LGTM

danyill commented:
Thank you for this detailed review and suggestions.
